### ISSUE TITLE
`manual_slice_fill` wrongly suggest non-const initializers

### ISSUE THREAD

**Issue by samueltardieu** [rust-lang/rust-clippy#14192]
### Summary

`manual_slice_fill` does not check:

- that the loop variable is used to index the slice being initialized
- that the expression doesn't reference the slice content (which may have been modified earlier in the loop) or the index (which is modified for every element)

### Lint Name

manual_slice_fill

### Reproducer

I tried this code:

```rust
fn main() {
    let mut tmp = vec![0; 3];

    for i in 0..tmp.len() {
        tmp[i] = i;
    }

    for i in 0..tmp.len() {
        tmp[0] = i;
    }
}
```

I saw this happen, for both loops:

```
warning: manually filling a slice
  --> /tmp/t.rs:8:5
   |
8  | /     for i in 0..tmp.len() {
9  | |         tmp[0] = i;
10 | |     }
   | |_____^ help: try: `tmp.fill(i);`
```

I expected to see this happen: no lint, since in the first loop the initializer depends on the loop variable, and in the second loop the loop variable is not used to index the slice.

### Version

```text
rustc 1.86.0-nightly (a9730c3b5 2025-02-05)
binary: rustc
commit-hash: a9730c3b5f84a001c052c60c97ed0765e9ceac04
commit-date: 2025-02-05
host: x86_64-unknown-linux-gnu
release: 1.86.0-nightly
LLVM version: 19.1.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by samueltardieu** [comment]
Ping @lapla-cogito

**Comment by lapla-cogito** [comment]
Thank you for letting me know, I submitted #14193.

**Comment by marxin** [comment]
Don't know how it's possible, but it seems this issue somehow sneaked into `1.86.0`:

```
❯ cargo +stable clippy
warning: manually filling a slice
   --> linker-utils/src/elf.rs:719:9
    |
719 | /         for b in &mut mask {
720 | |             *b = !*b;
721 | |         }
    | |_________^ help: try: `mask.fill(!*b);`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_slice_fill
    = note: `#[warn(clippy::manual_slice_fill)]` on by default
```

CI run: https://github.com/davidlattimore/wild/actions/runs/14258683395/job/39965815664?pr=642

**Comment by y21** [comment]
> Don't know how it's possible, but it seems this issue somehow sneaked into 1.86.0:

1.86 branched from master on Feb 14, and the fix only made it into the rust repository on Mar 2, so it missed the 1.86 release. It should be fixed on beta and will be in the next release 1.87.

**Comment by marxin** [comment]
I can confirm it's OK with the current nightly. I think the issue is pretty common, and I would consider backporting the fix to `1.86.1`, what do you think?
